packages/codecs/src/known/babe.ts

```diff
@@ -0,0 +1,157 @@
+  slotDuration: $.u64,
+
+  // The duration of epochs in slots.
+  epochLength: $.u64,
+
+  // A constant value that is used in the threshold calculation formula.
+  // Expressed as a rational where the first member of the tuple is the
+  // numerator and the second is the denominator. The rational should
+  // represent a value between 0 and 1.
+  // In the threshold formula calculation, `1 - c` represents the probability
+  // of a slot being empty.
+  c: $.Tuple($.u64, $.u64),
+
+  // The authorities for the genesis epoch.
+  authorities: $.Vec($.Tuple($AccountId32, $BabeAuthorityWeight)),
+
+  // The randomness for the genesis epoch.
+  randomness: $Randomness,
+
+  // Whether this chain should run with secondary slots, which are assigned
+  // in round-robin manner.
+  secondarySlots: $.bool,
+});
+
+export type BabeConfigurationV1 = $.Input<typeof $BabeConfigurationV1>;
+
+/*
+ * Unit type wrapper that represents a slot.
+ */
+export const $Slot = $.u64;
+
+export type Slot = $.Input<typeof $Slot>;
+
+/*
+ * Configuration data used by the BABE consensus engine that may change with epochs.
+ */
+export const $BabeEpochConfiguration = $.Struct({
+  // A constant value that is used in the threshold calculation formula.
+  // Expressed as a rational where the first member of the tuple is the
+  // numerator and the second is the denominator. The rational should
+  // represent a value between 0 and 1.
+  // In the threshold formula calculation, `1 - c` represents the probability
+  // of a slot being empty.
+  c: $.Tuple($.u64, $.u64),
+  // Whether this chain should run with secondary slots, which are assigned
+  // in round-robin manner.
+  allowedSlots: $AllowedSlots,
+});
+
+/*
+ * BABE epoch information
+ */
+export const $Epoch = $.Struct({
+  // The epoch index
+  epochIndex: $.u64,
+  // The starting slot of the epoch.
+  startSlot: $Slot,
+  // The duration of this epoch.
+  duration: $.u64,
+  // The authorities and their weights.
+  authorities: $.Vec($.Tuple($AccountId32, $BabeAuthorityWeight)),
+  // Randomness for this epoch.
+  randomness: $Randomness,
+  // Configuration of the epoch.
+  config: $BabeEpochConfiguration,
+});
+
+export type Epoch = $.Input<typeof $Epoch>;
+
+/*
+ * An opaque type used to represent the key ownership proof at the runtime API
+ * boundary. The inner value is an encoded representation of the actual key
+ * ownership proof which will be parameterized when defining the runtime. At
+ * the runtime API boundary this type is unknown and as such we keep this
+ * opaque representation, implementors of the runtime API will have to make
+ * sure that all usages of `OpaqueKeyOwnershipProof` refer to the same type.
+ */
+export const $OpaqueKeyOwnershipProof = $.PrefixedHex;
+
+export type OpaqueKeyOwnershipProof = $.Input<typeof $OpaqueKeyOwnershipProof>;
+
+/*
+ * An equivocation proof for multiple block authorships on the same slot (i.e. double vote).
+ */
+export const $EquivocationProof = $.Struct({
```

Comment:
Each consensus protocol might have its on `EquivocationProof`, so we probably need to add a prefix for codecs that potential have the same name. E.g: We also have a `EquivocationProof` in `grandpa` module as well. Similarly for the `Epoch` codec above.
```suggestion
export const $BabeEquivocationProof = $.Struct({
```
This might not be a problem for now, but the naming conflict might occur if we add runtime apis for other consensus module.